File: frontend/src/types/job.ts
/**
 * Job types.
 *
 * These mirror the response schemas in `backend/app/schemas/job.py`. The
 * backend serialises camelCase, so nothing is renamed here.
 *
 * Storage paths are deliberately absent: the backend never sends them, and the
 * client has no business knowing where a file lives on the server.
 */

import type { ErrorCode } from './api'

export type JobStatus = 'queued' | 'processing' | 'completed' | 'failed' | 'cancelled'

export type JobStage =
  | 'validating'
  | 'loading_model'
  | 'preprocessing'
  | 'running_inference'
  | 'postprocessing'
  | 'encoding'

export type OutputFormat = 'png' | 'jpeg' | 'webp'

/** Statuses a job can no longer move out of on its own. */
export const TERMINAL_STATUSES: readonly JobStatus[] = ['completed', 'failed', 'cancelled']

export function isTerminal(status: JobStatus): boolean {
  return TERMINAL_STATUSES.includes(status)
}

export interface ImageFacts {
  width: number
  height: number
  sizeBytes: number
  format: string
}

export interface JobError {
  code: ErrorCode
  detail: string
  technical?: string | null
}

export interface JobRecord {
  jobId: string
  status: JobStatus
  /** Null outside `processing`. */
  stage: JobStage | null
  progress: number
  model: string
  scale: number
  /** Null for a job submitted before modes existed, or without one. */
  mode: EnhancementMode | null
  /** Whether the size was asked for as a factor or as a destination. */
  outputType: OutputType
  /** The preset, when `outputType` is `target`. */
  target: TargetResolution | null
  /** Where it ran. Null until the job starts. */
  device: string | null
  input: ImageFacts
  output: ImageFacts | null
  /** Measured wall time, excluding queue wait. */
  processingMs: number | null
  error: JobError | null
  createdAt: string
  startedAt: string | null
  finishedAt: string | null
}

/** One page of history, as returned by `GET /api/jobs`. */
export interface JobPage {
  items: JobRecord[]
  /** Everything matching the filter, not just this page, so a pager can be sized. */
  total: number
  limit: number
  offset: number
}

export interface JobCreated {
  jobId: string
  status: JobStatus
  /** 0 means it is next, or already started. */
  queuePosition: number
  createdAt: string
}

/** The optional `settings` object of a submission. */
export interface EnhanceSettings {
  sharpenStrength?: number
  denoiseStrength?: number
  tileSize?: number | null
  tilePad?: number | null
}

/**
 * What the enhancement should prioritise.
 *
 * A named bundle of decisions the backend makes - which model, how much
 * denoising - so the UI never has to name a model to express an intent.
 */
export type EnhancementMode = 'standard' | 'creative'

/**
 * A requested output size, named by its long edge.
 *
 * Deliberately not an upscale factor. "4K" is a destination the output lands
 * on whatever the input was; "4x" is a multiplier. The two are different
 * questions and the UI keeps them apart.
 *
 * 16K makes that impossible to miss: it is eight times the 1920 the family is
 * built on, so a 1080p source reaches it with an 8x pass, not a 16x one.
 */
export type TargetResolution = '2k' | '4k' | '6k' | '8k' | '16k'

/**
 * How a job's output size was asked for.
 *
 * A job from before target resolutions existed carries no metadata and reads
 * as `scale`, which is what it was - that was the only way to ask.
 */
export type OutputType = 'scale' | 'target'

export const TARGET_LONG_EDGE: Record<TargetResolution, number> = {
  '2k': 1920,
  '4k': 3840,
  '6k': 5760,
  '8k': 7680,
  '16k': 15360,
}

/**
 * The model each mode runs when the user has not named one.
 *
 * A mirror of `STANDARD_MODEL` and `CREATIVE_MODEL` in
 * `backend/app/services/mode_planner.py`, and only a mirror: **the backend
 * remains authoritative**. It resolves the model itself and ignores anything
 * the client believes, so nothing here can change which weights run.
 *
 * What it is for is the opposite problem. The panel has to show a model before
 * the user picks one, and it used to show whichever model happened to be
 * adopted first - so Creative displayed `RealESRGAN_x4plus`, read
 * `supportsDenoise` off it, and disabled the noise-reduction slider, while the
 * job ran `realesr-general-x4v3` at denoise 0.25. Three statements on screen,
 * all wrong for the job that ran, and the one control Creative exists to expose
 * greyed out.
 *
 * A contract test on the backend reads this table out of the TypeScript source
 * and asserts the two agree, the same way it already does for
 * `TARGET_LONG_EDGE`, so a mode whose model changes on the server cannot drift
 * from what the browser shows.
 */
export const MODE_DEFAULT_MODEL: Record<EnhancementMode, string> = {
  standard: 'RealESRGAN_x4plus',
  creative: 'realesr-general-x4v3',
}

export interface CreateJobRequest {
  file: File
  /**
   * Omitted when the user has not chosen a model, which lets `mode` supply one.
   * The backend already accepts an absent model and falls back to the mode's
   * default and then to `default_model`.
   */
  model?: string
  /** Ignored when `target` is set - the two answer the same question. */
  scale: number
  format: OutputFormat
  quality?: number
  preserveMetadata: boolean
  settings: EnhanceSettings
  mode?: EnhancementMode
  /** When set, the backend plans the factor and resizes to this exactly. */
  target?: TargetResolution
}

/**
 * What the progress stream sends. Names match the SSE `event:` field.
 *
 * `tilesDone`/`tilesTotal` appear only during inference, because that is the
 * only stage with a real count behind it — the backend does not invent one for
 * the others, and neither does this type.
 */
export interface ProgressEventData {
  progress: number
  stage: JobStage
  tilesDone?: number
  tilesTotal?: number
}

export interface StageEventData {
  stage: JobStage
  progress: number
}

export interface CompletedEventData {
  jobId: string
  processingMs: number | null
}

export interface FailedEventData {
  code: ErrorCode
  detail: string
  technical?: string
}

export interface CancelledEventData {
  jobId: string
}

/**
 * The `job:stage.*` key naming a stage. The caller translates it.
 *
 * A key rather than a label, so the stage a job reports over SSE stays a
 * protocol value and only becomes words at the point it is shown.
 */
export function stageKey(stage: JobStage): string {
  return `stage.${stage}`
}

File: backend/app/schemas/job.py
"""Job request and response schemas.

Mirrors the contract in docs/api.md. Storage paths are deliberately absent from
every model here: they live in the database and never in a response body
(architecture § 10).
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.models.db import Job
from app.models.enums import (
    EnhancementMode,
    JobStage,
    JobStatus,
    OutputFormat,
    OutputType,
    TargetResolution,
)
from app.repositories.base import Page
from app.schemas.common import CamelModel


class EnhanceSettings(CamelModel):
    """The optional `settings` object of a submission."""

    sharpen_strength: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Unsharp-mask post-process. Not an AI feature, and not labelled as one.",
    )
    denoise_strength: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description=(
            "DNI interpolation coefficient, for models with supportsDenoise. "
            "1.0 denoises most; 0.0 preserves noise. "
            "Omitting the field is not the same as sending 0: the server "
            "resolves an omitted value to the mode's setting, or to 0.25 when "
            "no mode names one, and only ever to null for a model with no "
            "denoise pair. Send an explicit value to override that."
        ),
    )
    tile_size: int | None = Field(
        default=None,
        ge=0,
        le=2048,
        description="Override TILE_SIZE. Null decides automatically from free VRAM.",
    )
    tile_pad: int | None = Field(default=None, ge=0, le=128, description="Override TILE_PAD.")


class JobCreatedResponse(CamelModel):
    """202 Accepted: the job exists and is waiting for a worker."""

    job_id: str
    status: JobStatus
    queue_position: int = Field(description="0 means it is next, or already started")
    created_at: datetime


class ImageFacts(CamelModel):
    """Dimensions and weight of an image, input or output."""

    width: int
    height: int
    size_bytes: int
    format: str


class JobError(CamelModel):
    """Why a job failed, in the same shape as a problem document."""

    code: str
    detail: str
    technical: str | None = None


def _describe_request(
    job: Job,
) -> tuple[EnhancementMode | None, OutputType, TargetResolution | None]:
    """Read how the job was asked for out of its stored options.

    All of this lives in the `enhance_options` JSON, which exists so an option
    can be added without a migration. A job submitted before modes existed
    carries none of it and reads as an unnamed mode producing a scale-based
    output - which is exactly what it was.

    Unknown values are treated as absent rather than raised on: a row written
    by a newer build should not make an older one unable to list its history.
    """
    options = job.enhance_options or {}

    mode: EnhancementMode | None = None
    raw_mode = options.get("mode")
    if isinstance(raw_mode, str) and raw_mode in set(EnhancementMode):
        mode = EnhancementMode(raw_mode)

    target: TargetResolution | None = None
    raw_target = options.get("target")
    if isinstance(raw_target, str) and raw_target in set(TargetResolution):
        target = TargetResolution(raw_target)

    # A target is only claimed when the preset survived; a job with the pixel
    # dimensions but no recognisable preset is still a target job, so the
    # dimensions are the fallback signal.
    has_target = target is not None or options.get("targetWidth") is not None
    output_type = OutputType.TARGET if has_target else OutputType.SCALE

    return mode, output_type, target


class JobResponse(CamelModel):
    """The full job record."""

    job_id: str
    status: JobStatus
    stage: JobStage | None = Field(default=None, description="Null outside processing")
    progress: int = Field(ge=0, le=100)
    model: str
    scale: int
    mode: EnhancementMode | None = Field(
        default=None,
        description="Null for a job submitted before modes existed, or without one.",
    )
    output_type: OutputType = Field(
        description="Whether the size was asked for as a factor or as a destination."
    )
    target: TargetResolution | None = Field(
        default=None, description="The preset, when output_type is target."
    )
    device: str | None = Field(default=None, description="Where it ran; null until it starts")
    input: ImageFacts
    output: ImageFacts | None = None
    processing_ms: int | None = Field(
        default=None, description="Measured wall time, excluding queue wait"
    )
    error: JobError | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None

    @classmethod
    def from_job(cls, job: Job) -> JobResponse:
        """Build the response from a row, leaving every path behind."""
        status = JobStatus(job.status)
        mode, output_type, target = _describe_request(job)

        output: ImageFacts | None = None
        if status is JobStatus.COMPLETED and job.output_width and job.output_height:
            output = ImageFacts(
                width=job.output_width,
                height=job.output_height,
                size_bytes=job.output_bytes or 0,
                format=OutputFormat(job.output_format).value.upper(),
            )

        error: JobError | None = None
        if job.error_code is not None:
            error = JobError(
                code=job.error_code,
                detail=job.error_message or "The job failed.",
                technical=job.error_technical,
            )

        return cls(
            job_id=job.id,
            status=status,
            mode=mode,
            output_type=output_type,
            target=target,
            stage=JobStage(job.stage) if job.stage else None,
            progress=job.progress,
            model=job.model_name,
            scale=job.scale,
            device=job.device,
            input=ImageFacts(
                width=job.input_width,
                height=job.input_height,
                size_bytes=job.input_bytes,
                format=job.input_format,
            ),
            output=output,
            processing_ms=job.processing_ms,
            error=error,
            created_at=job.created_at,
            started_at=job.started_at,
            finished_at=job.finished_at,
        )


class JobPage(CamelModel):
    """One page of history.

    `total` is the size of the whole filtered set, not of this page, so a
    client can show "20 of 42" and size its pagination without a second call.
    """

    items: list[JobResponse]
    total: int = Field(description="Matching jobs in total, ignoring limit and offset")
    limit: int
    offset: int

    @classmethod
    def from_page(cls, page: Page) -> JobPage:
        return cls(
            items=[JobResponse.from_job(job) for job in page.items],
            total=page.total,
            limit=page.limit,
            offset=page.offset,
        )
